test/playwright/visual.spec.js

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (C) 2020-2022  Yomichan Authors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+const path = require('path');
+const {test: base, chromium} = require('@playwright/test');
+const root = path.join(__dirname, '../../');
+
+export const test = base.extend({
+    context: async ({ }, use) => { // eslint-disable-line
+        const pathToExtension = path.join(root, 'ext');
+        const context = await chromium.launchPersistentContext('', {
+            // headless: false,
+            args: [
+                '--headless=new',
+                `--disable-extensions-except=${pathToExtension}`,
+                `--load-extension=${pathToExtension}`
+            ]
+        });
+        await use(context);
+        await context.close();
+    },
+    extensionId: async ({context}, use) => {
+        let [background] = context.serviceWorkers();
+        if (!background) {
+            background = await context.waitForEvent('serviceworker');
+        }
+
+        const extensionId = background.url().split('/')[2];
+        await use(extensionId);
+    }
+});
+const expect = test.expect;
+
+test('visual', async ({page, extensionId}) => {
+    // Install jmdict_english.zip
+    await page.goto(`chrome-extension://${extensionId}/settings.html`);
+    await page.waitForTimeout(500); // wait for welcome.html page to load, so we can then bringToFront() -- otherwise, the screenshot can hang
+    await expect(page.locator('id=dictionaries')).toBeVisible();
+    await page.bringToFront(); // improves performance
+    await expect.soft(page).toHaveScreenshot('settings-fresh.png', {maxDiffPixels: 100}); // maxDiffPixels is needed because the disk usage indicator is different per run
+    await page.waitForTimeout(200);
+    await page.locator('input[id="dictionary-import-file-input"]').setInputFiles(path.join(root, 'dictionaries/jmdict_english.zip'));
+    await expect(page.locator('id=dictionaries')).toHaveText('Dictionaries (1 installed, 1 enabled)', {timeout: 5 * 60 * 1000});
+    await expect.soft(page).toHaveScreenshot('settings-jmdict-loaded.png', {maxDiffPixels: 100}); // maxDiffPixels is needed because the disk usage indicator is different per run
```

Comment:
I wonder why 100 pixels has been chosen as the threshold

Reply:
Arbitrarily. It could probably be more precise.